tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,388 @@
+        #   (prev_param - cur_param)/base_lr
+        momentum = (
+            slowmo_optim.slowmo_factor * initial_slow_momentum_buffer
+            + (initial_prev_params - fsdp_net.params[0]) / learning_rate
+        )
+
+        # parameter_(t+1) = prev_param - slowmo_lr * base_lr * momentum_(t+1)
+        calculated_params = initial_prev_params - 0.1 * learning_rate * momentum
+
+        self.assertEqual(fsdp_net_slowmo.params[0], calculated_params)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("sharding_strategy", [ShardingStrategy.NO_SHARD])
+    def test_slowmo_optimizer_state_dict(
+        self, sharding_strategy: Optional[ShardingStrategy]
+    ):
+        chkpt = tempfile.gettempdir() + "/checkpoint.pt"
+        fsdp_net_slowmo = FSDP(
+            Net(has_wrapping=False, sharding_strategy=sharding_strategy),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        ).to(self.rank)
+
+        cur_subgroup = dist.new_group(ranks=[self.rank])
+        slowmo_state = slowmo_comm.SlowMoState(cur_subgroup)
+        fsdp_net_slowmo.register_comm_hook(slowmo_state, slowmo_comm.slowmo_hook)
+        inpt = torch.randn(  # type: ignore[call-overload]
+            (7, 8), dtype=torch.float, device=self.rank
+        )
+
+        slowmo_optim = self._init_slowmo_optimizer(
+            base_optim=torch.optim.SGD(fsdp_net_slowmo.parameters(), lr=1e-2),
+            slowmo_freq=4,
+        )
+
+        for _ in range(10):
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim)
+
+        state = {"optim_state_dict": slowmo_optim.state_dict()}
+
+        if self.rank == 0:
+            torch.save(state, chkpt)
+
+        dist.barrier()
+
+        map_location = {"cuda:%d" % 0: "cuda:%d" % self.rank}
+        checkpoint = torch.load(chkpt, map_location=map_location)
+
+        # initialize dummy optimizer with different parameters
+        slowmo_optim_dummy = slowmo_optimizer.SlowMomentumOptimizer(
+            base_optim=torch.optim.SGD(fsdp_net_slowmo.parameters(), lr=1e-2),
+            slowmo_freq=2,
+            slowmo_factor=3,
+            slowmo_lr=4,
+        )
+        slowmo_optim_dummy.load_state_dict(checkpoint["optim_state_dict"])
+
+        # make sure averager's period and step were updated
+        self.assertEqual(
+            slowmo_optim_dummy.averager.period, slowmo_optim.averager.period
+        )
+        self.assertEqual(slowmo_optim_dummy.averager.step, slowmo_optim.averager.step)
+
+        # make sure slowmo parameters were updated
+        self.assertEqual(slowmo_optim_dummy.slowmo_freq, slowmo_optim.slowmo_freq)
+        self.assertEqual(slowmo_optim_dummy.slowmo_factor, slowmo_optim.slowmo_factor)
+        self.assertEqual(slowmo_optim_dummy.slowmo_lr, slowmo_optim.slowmo_lr)
+
+        for _ in range(10):
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim_dummy)
+
+        self.assertEqual(slowmo_optim_dummy.averager.step, 20)
```

Comment:
Should we refactor 10 into a variable if it is related to 20 (e.g. 20 is supposed to 2 times the number of iterations)? (I did not verify if that is indeed the case.)